--- gstreamer-vaapi-0.6.0/configure.ac.orig	2015-07-15 14:49:33.000000000 +0200
+++ gstreamer-vaapi-0.6.0/configure.ac	2015-08-03 21:09:01.565451502 +0200
@@ -167,6 +167,7 @@
 GLIB_VERSION_REQUIRED=glib_version
 PKG_CHECK_MODULES([GLIB], [glib-2.0 >= $GLIB_VERSION_REQUIRED])
 AC_SUBST(GLIB_VERSION_REQUIRED)
+PKG_CHECK_MODULES([GMODULE], [gmodule-2.0])
 
 dnl Check to see if dlopen is in default libraries (like Solaris, which
 dnl has it in libc), or if libdl is needed to get it.
--- gstreamer-vaapi-0.6.0/gst-libs/gst/vaapi/Makefile.am.orig	2015-07-15 14:47:43.000000000 +0200
+++ gstreamer-vaapi-0.6.0/gst-libs/gst/vaapi/Makefile.am	2015-08-03 21:09:17.892117476 +0200
@@ -502,6 +502,7 @@
 
 libgstvaapi_egl_@GST_API_VERSION@_la_LIBADD =	\
 	$(GLIB_LIBS)				\
+	$(GMODULE_LIBS)				\
 	$(GST_LIBS)				\
 	$(GST_BASE_LIBS)			\
 	$(GST_VIDEO_LIBS)			\
